x/protorev/keeper/posthandler.go

```diff
@@ -193,3 +201,49 @@ func extractSwapOutPools(routes []poolmanagertypes.SwapAmountOutRoute, tokenOutD
 
 	return swappedPools
 }
+
+// extractJoinPools extracts the pools that were swapped on for a MsgJoinSwapExternAmountIn or MsgJoinSwapShareAmountOut
+func (k Keeper) extractJoinPools(ctx sdk.Context, poolId uint64, tokenInDenom string) []SwapToBackrun {
+	swappedPools := make([]SwapToBackrun, 0)
+	pool, err := k.gammKeeper.GetPoolAndPoke(ctx, poolId)
+	if err != nil {
+		return swappedPools
+	}
+	// Get all the pool coins and iterate to get the denoms that make up the swap
+	coins := pool.GetTotalPoolLiquidity(ctx)
+	for _, coin := range coins {
+		if coin.Denom == tokenInDenom {
+			continue
+		}
+
+		swappedPools = append(swappedPools, SwapToBackrun{
+			PoolId:        poolId,
+			TokenOutDenom: coin.Denom,
+			TokenInDenom:  tokenInDenom})
+	}
+
+	return swappedPools
+}
+
+// extractExitPools extracts the pools that were swapped on for a MsgExitSwapExternAmountOut or MsgExitSwapShareAmountIn
+func (k Keeper) extractExitPools(ctx sdk.Context, poolId uint64, tokenOutDenom string) []SwapToBackrun {
```

Comment:
These two functions are basically identical. You could copy a lot less code by changing the abstraction, so the caller indicates whether tokenIn or tokenOut denom is being passed.

Reply:
Changed in: 22248313b1f2664e02e64a48607c1d3e1f36e83e